client/webserver/site/src/js/mmsettings.ts

```diff
@@ -633,261 +1177,418 @@ export default class MarketMakerSettingsPage extends BasePage {
-          this.baseWalletSettingControl[opt].setValue(value)
-        }
+    // Buy/Sell placements
+    oldCfg.buyPlacements.forEach((p) => { this.addPlacement(true, p, viewOnly) })
+    oldCfg.sellPlacements.forEach((p) => { this.addPlacement(false, p, viewOnly) })
+
+    for (const opt of Object.keys(cfg.baseOptions)) {
+      const value = oldCfg.baseOptions[opt]
+      cfg.baseOptions[opt] = value
+      if (this.baseWalletSettingControl[opt]) {
+        this.baseWalletSettingControl[opt].setValue(value)
       }
     }
 
     // Quote wallet options
-    if (this.updatedConfig.basicMarketMakingConfig.quoteOptions && this.originalConfig.basicMarketMakingConfig.quoteOptions) {
-      for (const opt of Object.keys(this.updatedConfig.basicMarketMakingConfig.quoteOptions)) {
-        const value = this.originalConfig.basicMarketMakingConfig.quoteOptions[opt]
-        this.updatedConfig.basicMarketMakingConfig.quoteOptions[opt] = value
-        if (this.quoteWalletSettingControl[opt]) {
-          this.quoteWalletSettingControl[opt].setValue(value)
-        }
+    for (const opt of Object.keys(cfg.quoteOptions)) {
+      const value = oldCfg.quoteOptions[opt]
+      cfg.quoteOptions[opt] = value
+      if (this.quoteWalletSettingControl[opt]) {
+        this.quoteWalletSettingControl[opt].setValue(value)
       }
     }
 
     this.updateModifiedMarkers()
   }
 
+  /*
+   * validateFields validates configuration values and optionally shows error
+   * messages.
+   */
+  validateFields (showErrors: boolean): boolean {
+    let ok = true
+    const {
+      page, specs: { botType },
+      updatedConfig: { sellPlacements, buyPlacements, profit, useEmptyMarketRate, emptyMarketRate, baseBalance, quoteBalance }
+    } = this
+    const setError = (errEl: PageElement, errID: string) => {
+      ok = false
+      if (!showErrors) return
+      errEl.textContent = intl.prep(errID)
+      Doc.show(errEl)
+    }
+    if (showErrors) {
+      Doc.hide(
+        page.buyPlacementsErr, page.sellPlacementsErr, page.profitInputErr,
+        page.emptyMarketRateErr, page.baseBalanceErr, page.quoteBalanceErr
+      )
+    }
+    if (botType !== botTypeBasicArb) {
+      if (buyPlacements.length === 0) setError(page.buyPlacementsErr, intl.ID_NO_PLACEMENTS)
+      if (sellPlacements.length === 0) setError(page.sellPlacementsErr, intl.ID_NO_PLACEMENTS)
+    }
+    if (botType !== botTypeBasicMM) {
+      if (isNaN(profit)) setError(page.profitInputErr, intl.ID_INVALID_VALUE)
+      else if (profit === 0) setError(page.profitInputErr, intl.ID_NO_ZERO)
+    } else { // basic mm
+      // TODO: Should we enforce an empty market rate if there are no
+      // oracles?
+      if (useEmptyMarketRate && emptyMarketRate === 0) setError(page.emptyMarketRateErr, intl.ID_NO_ZERO)
+      if (baseBalance === 0) setError(page.baseBalanceErr, intl.ID_NO_ZERO)
+      if (quoteBalance === 0) setError(page.quoteBalanceErr, intl.ID_NO_ZERO)
```

Comment:
Should we allow empty balance for one of the assets? They could start off with 0, then after the first trade have balance on that asset.

Reply:
Yeah. Removing.

I think there are some additional validation steps that can be done during configuration, but it'll be a project of its own. I plan to follow up with some "quick configuration" options that will help the user get their settings right without having to manually adjust everything.